rfc/001-composable-pipelines/README.md

```diff
@@ -0,0 +1,345 @@
+
+- **Asynchronous Execution**: pipelines fundamentally support asynchronous
+  execution by sequentializing dataflow within an operator, and parallelizing it
+  across operators. This maps well to the actor model architecture of VAST.
+  Operators naturally map to actors.
+
+  As an optimization, an executor may place multiple operators in the same
+  sequential execution to avoid asynchronous context switches.
+
+- **Strong Typing**: pipelines are strongly typed with respect to their input
+  and output types. This enables static type checking when instantiating a
+  pipeline, e.g., an output type of `Arrow` with a downstream operator that has
+  `Bytes` as input type describes an invalid pipeline.
+
+  At the data plane, within Arrow, a "type" consists of a specific schema of a
+  record batch. We should consider a mechanism to perform cross-operator type
+  checking to simplify the implementation of individual operators.
+
+- **Push vs. Pull**: implementation of a dataflow system can follow push-based
+  execution where a passive pipeline only does work when some component feeds it
+  with data, which then triggers a cascade of executions from the opt. The
+  alternative is pull-based, where consumers request data at the sink. This
+  request triggers request for additional results upstream, interleaved with
+  computation to produce them. [Combining both models][shaikhha18] has shown
+  promising results.
+
+  Ideally we can find an API that allows for both execution modes so that we can
+  optimize execution incrementally, without locking ourselves into one corner.
+
+[shaikhha18]: https://doi.org/10.1017/s0956796818000102
+
+## Alternatives
+
+We considered the following ideas that we do not want to pursue further.
+
+### Pipelines are only a query language construct
+
+We may consider pipelines only in the context of executing queries at the
+server. After all, this is where we are going to implement them first. However,
+this limits us unnecessarily, given that pipeline can already be deployed a
+various points in VAST today. We must consider them as general building block
+that is easy to instantiate in various contexts.
+
+### Commands are pipeline operators
+
+To illustrate the composability of pipelines, we could make every VAST command
+an operator:
+
+```bash
+vast from s3://aws |
+  vast read json |
+  vast 'summarize count(dst) group-by src' |
+  vast write feather |
+  vast to /path/to/file.feather
+```
+
+As there is no clear underlying use case outside of testing and debugging, we
+are not going to discuss this idea in more depth.
+
+### Statically typed pipelines
+
+We could implement the entire pipeline with Arrow IPC streams in the "interior"
+operators, with only `from` and `to` taking raw input as byte stream. Such
+static typing may be too restrictive, preventing a wide array of operators.
+We therefore assume that input and output are dynamically typed.
```

Comment:
I would just remove this again. We don't know this for certain without any actual implementation experience. As-is, this is just a guess, and that doesn't belong into a proposal.